raft/raft.go

```diff
@@ -423,10 +475,59 @@ func (r *raft) send(m pb.Message) {
 			m.Term = r.Term
 		}
 	}
-	if m.To == r.id {
-		r.logger.Panicf("message should not be self-addressed when sending %s", m.Type)
+	if m.Type == pb.MsgAppResp || m.Type == pb.MsgVoteResp || m.Type == pb.MsgPreVoteResp {
```

Comment:
Pre-votes are not persisted (right? it's been a long time), so I don't think MsgPreVoteResp needs to be here.

Reply:
Pre-votes aren't persisted, but they are based on the state of the node's log (see `r.raftLog.isUpToDate`). By including MsgPreVoteResp here, we ensure that the response is derived from stable state.

Given pre-vote's optional role in the Raft algorithm and the limited amount of state that nodes record from MsgPreVoteResp's they receive from others, I can't think of a case where responding based on unstable state would cause an issue. However, I still think it's best to keep it here unless we find that it's important not to and then go through the effort of proving its correctness.

That's similar to why I avoided adding an optimization for message rejections (see the last paragraph in this comment, which starts with `Rejected responses`).